django/conf/__init__.py

```diff
@@ -239,6 +266,33 @@ def __init__(self, settings_module):
         if self.is_overridden("USE_L10N"):
             warnings.warn(USE_L10N_DEPRECATED_MSG, RemovedInDjango50Warning)
 
+        self.STORAGES = {}
+
+        if self.is_overridden("DEFAULT_FILE_STORAGE"):
+            if self.is_overridden("STORAGES"):
+                raise ImproperlyConfigured(STORAGES_CONFLICT_MESSAGE)
+            self.STORAGES[DEFAULT_STORAGE_ALIAS] = {
+                "BACKEND": self.DEFAULT_FILE_STORAGE
+            }
+            warnings.warn(DEFAULT_FILE_STORAGE_DEPRECATED_MSG, RemovedInDjango51Warning)
+
+        if self.is_overridden("STATICFILES_STORAGE"):
+            if self.is_overridden("STORAGES"):
+                raise ImproperlyConfigured(STORAGES_CONFLICT_MESSAGE)
+            self.STORAGES[STATICFILES_STORAGE_ALIAS] = {
+                "BACKEND": self.STATICFILES_STORAGE
+            }
+            warnings.warn(STATICFILES_STORAGE_DEPRECATED_MSG, RemovedInDjango51Warning)
+
+        self.STORAGES.setdefault(
+            DEFAULT_STORAGE_ALIAS,
+            {"BACKEND": "django.core.files.storage.FileSystemStorage"},
+        )
+        self.STORAGES.setdefault(
+            STATICFILES_STORAGE_ALIAS,
+            {"BACKEND": "django.contrib.staticfiles.storage.StaticFilesStorage"},
+        )
```

Comment:
We should add `STORAGES` to the `django/conf/global_settings.py` instead (with hard-coded values):
```python
STORAGES = {
    "default": {
        {"BACKEND": "django.core.files.storage.FileSystemStorage"}
    },
    "staticfiles": {
        {"BACKEND": "django.contrib.staticfiles.storage.StaticFilesStorage"},
    },
}
```

Reply:
It's more important to make a removal straightforward than keeping this DRY.

Reply:
Moving `STORAGES` to `django/conf/global_settings.py` breaks tests because a reference to this dict is shared between tests.
You can reproduce it by removing `self.STORAGES = {}` from `Settings.__init__`, adding `STORAGES = {}` to global settings and running:
`python runtests.py deprecation forms_tests --parallel 1`

Reply:
This is an isolation issue that should be fixed. `StorageHandler` cannot share the same global setting, you can check how it's handled in `BaseConnectionHandler` via `configure_settings()`.

Reply:
Overriding `STORAGES` setting in settings and tests could require setting both the default and staticfiles storages if this setting is hard coded. Using `setdefault` here can save users some boilerplate code.
Say I want to use a custom default file storage and a default staticfiles storage in a test. With `setdefault` I can rely on the default staticfiles storage without explicitly defining it in the `override_settings` decorator. Having sensible defaults is a better developer experience.

Removal of the deprecated settings won't require the removal of the bits with `setdefault` so I don't see how they would complicate the deprecation process.

This isolation issue with a common dict in `django/conf/global_settings.py` is caused by the deprecated `DEFAULT_FILE_STORAGE` and `STATICFILES_STORAGE` settings as they modify the new `STORAGES` setting for backward compatibility.
`BaseConnectionHandler` doesn't need to handle this corner case in `configure_settings()` because it doesn't support some deprecated settings that modify `DATABASES` dict (imagine deprecating something like `SQLITE_DATABASE_PATH`).
I could add `STORAGES = {}` to `django/conf/global_settings.py` and change `self.STORAGES = {}` to `self.STORAGES = self.STORAGES.copy()` here but I don't see a point.
I could also move handling of the deprecated settings to the `StorageHandler` but it would require to use them outside of `django/conf/__init__.py` which will make the removal less straightforward.

Reply:
@felixxm Is the current implementation acceptable?

Reply:
I don't think we can special-case `STORAGES` here. I think it should be added to `django/conf/global_settings.py` as per https://github.com/django/django/pull/15610/files#r946605556 

(It's a much bigger proposal to adjust how `Settings` works.)

Reply:
Ok, I'll add `STORAGES` to `django/conf/global_settings.py`.

Can I call `STORAGES.copy()` in `Settings.__init__` or is it special-casing STORAGES you've mentioned?

Alternatively I can handle deprecated settings in `StorageHandler`. In this case we'll have to move them outside of `django/conf/__init__.py` which will make the removal less straightforward.

Reply:
If using is necessary in order to allow the deprecation then that's going to have to be, with an appropriate `# RemovedIn...` comment.

Reply:
Fixed